libc/getpw.c

```diff
@@ -0,0 +1,111 @@
+/*
+ * Phoenix-RTOS
+ *
+ * libc-tests
+ *
+ * Testing getpwuid and getpwnam edge cases
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Mateusz Niewiadomski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+/* make compilable against glibc (all these tests were run on host against libc/linux kernel) */
+#define _GNU_SOURCE
+
+#include <pwd.h>
+#include <unistd.h>
+
+#include <unity_fixture.h>
+
+
+static unsigned int ispasswdfile;
+static const char root[] = "root";
+static struct passwd *pw;
+
+
+TEST_GROUP(getpwd);
+
+
+TEST_SETUP(getpwd)
+{
+	ispasswdfile = (access("/etc/passwd", F_OK) == 0) ? 1 : 0;
+	pw = NULL;
+}
+
+
+TEST_TEAR_DOWN(getpwd)
+{
+}
+
+/* Get root account details by name */
+TEST(getpwd, getpwnam_getroot)
+{
+	pw = getpwnam(root);
+	if (ispasswdfile) {
+		TEST_ASSERT_NOT_NULL(pw);
+
+		TEST_ASSERT_EQUAL_STRING("root", pw->pw_name);
+		/* no password hash checking */
+		TEST_ASSERT_EQUAL_INT(0, pw->pw_uid);
+		TEST_ASSERT_EQUAL_INT(0, pw->pw_gid);
+		TEST_ASSERT_EQUAL_STRING("root", pw->pw_gecos);
+		TEST_ASSERT_EQUAL_STRING("/", pw->pw_dir);
+		TEST_ASSERT_EQUAL_STRING("/bin/sh", pw->pw_shell);
+	}
+	else {
+		TEST_ASSERT_NULL(pw);
```

Comment:
I don't like that we're marking this test as `PASSED` even if it doesn't check anything relevant.

Maybe we should try exiting with `TEST_IGNORE()` - it might be visible as a skipped one then (or don't run the test at all - as described above).

Reply:
I propose the middle path, as if there is no `/etc/passwd` file in the filesystem we should still check if the functions behave accordingly. So I propose to:
 - try to execute `getpwuid`/`getpwnam` no matter what
 - if there is no `/etc/passwd` check if `pw` is NULL
 - only if NULL check passes - ignore the test

Additionally add tests: `getpwuid_nopasswdfile` and `getpwnam_nopasswdfile` as this is valid test scenario